### PR TITLE
(fleet) add installer telemetry

### PR DESCRIPTION
**Pull request by arbll**
This PR introduces telemetry in the installer in the form of traces.
Those traces are sent to APM's existing telemetry infra.

This should provide an easy to extend way to ship both traces and logs
from the installer.

### CODE REVIEW

File: pkg/updater/updater.go [hunk 1]
```diff
@@ -244,16 +201,19 @@ func (u *updaterImpl) Start(_ context.Context) error {
...
 	u.requestsWG.Wait()
 	close(u.requests)
 	return nil
 }
 
 // Bootstrap installs the stable version of the package.
-func (u *updaterImpl) Bootstrap(ctx context.Context, pkg string) error {
+func (u *updaterImpl) BootstrapDefault(ctx context.Context, pkg string) (err error) {
+	span := tracer.StartSpan("BootstrapDefault")
+	defer func() { span.Finish(tracer.WithError(err)) }()
```
Comment (paullegranddc) [inline pkg/updater/updater.go:216]: Shouldn't you you use [StartSpanFromContext](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1@v1.62.0/ddtrace/tracer#StartSpanFromContext) to propagate the trace with the context?

File: pkg/internaltelemetry/client.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+// NewClient creates a new telemetry client
+func NewClient(httpClient httpClient, endpoints []*config.Endpoint, service string) Client {
+	info := metadatautils.GetInformation()
+	return &client{
+		client:    httpClient,
+		endpoints: endpoints,
+		baseEvent: Event{
+			APIVersion: "v2",
+			DebugFlag:  true,
```
Comment (paullegranddc) [inline pkg/internaltelemetry/client.go:116]: Could you make that a parameter so we don't forget to turn it off in the release?

File: pkg/updater/telemetry.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	client internaltelemetry.Client
+
+	site       string
+	socketPath string
+	listener   net.Listener
+	server     *http.Server
+}
+
+func newTelemetry(config config.Reader, socketDirectory string) (*telemetry, error) {
+	socketPath := filepath.Join(socketDirectory, "telemetry.sock")
```
Comment (raphaelgavache) [inline pkg/updater/telemetry.go:39]: a little comment explaining the internal socket hack to get go tracer traces would be nice for future readers

File: pkg/updater/updater.go [hunk 4]
```diff
@@ -214,7 +242,9 @@ func (u *updaterImpl) BootstrapVersion(ctx context.Context, pkg string, version
 }
 
 // BootstrapURL installs the stable version of the package.
-func (u *updaterImpl) BootstrapURL(ctx context.Context, url string) error {
+func (u *updaterImpl) BootstrapURL(ctx context.Context, url string) (err error) {
+	span, ctx := tracer.StartSpanFromContext(ctx, "BootstrapURL")
```
Comment (raphaelgavache) [inline pkg/updater/updater.go:246]: operation names are used in datadog metrics and resource as tags. By default operation name = resource
the metric is case sensitive and the tag is converted to a lower case

for nice interaction in the app, the easiest is to follow tags normalisation: https://docs.datadoghq.com/getting_started/tagging/?s=metric#define-tags

File: pkg/updater/updater.go [hunk 5]
```diff
@@ -248,15 +278,17 @@ func (u *updaterImpl) boostrapPackage(ctx context.Context, url string, expectedP
 }
 
 // StartExperiment starts an experiment with the given package.
-func (u *updaterImpl) StartExperiment(ctx context.Context, pkg string, version string) error {
+func (u *updaterImpl) StartExperiment(ctx context.Context, pkg string, version string) (err error) {
+	span, ctx := tracer.StartSpanFromContext(ctx, "StartExperiment")
```
Comment (raphaelgavache) [inline pkg/updater/updater.go:282]: reminder for the lower case on all spans

**Review by raphaelgavache — APPROVED** [review]
lgtm

**Review by derekwbrown — APPROVED** [review]
Not sure why we're owners on pkg/internaltelemetry ;).

would appreciate as part of the QA for the windows agent crash detection to be tested.

Instructions in this PR https://github.com/DataDog/datadog-agent/pull/23088

follow the instructions starting at "for additional testing"

**Pull request merged**